internal/controllers/mic_reconciler.go

```diff
@@ -0,0 +1,336 @@
+		errs = append(errs, err)
+	}
+
+	return errors.Join(errs...)
+}
+
+func (mrhi *micReconcilerHelperImpl) updateStatusByMBSC(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig) error {
+	mbsc, err := mrhi.mbscHelper.GetMBSC(ctx, micObj.Name, micObj.Namespace)
+	if err != nil {
+		return fmt.Errorf("failed to get ModuleBuildSignConfig object %s/%s: %v", micObj.Namespace, micObj.Name, err)
+	}
+
+	if mbsc == nil {
+		return nil
+	}
+
+	patchFrom := client.MergeFrom(micObj.DeepCopy())
+	for _, imageState := range mbsc.Status.ImagesStates {
+		imageSpec := mrhi.micHelper.GetModuleImageSpec(micObj, imageState.Image)
+		if imageSpec == nil {
+			// image not found in spec, ignore
+			continue
+		}
+		micImageStatus := kmmv1beta1.ImageExists
+		if imageState.Status == kmmv1beta1.ImageBuildFailed {
+			micImageStatus = kmmv1beta1.ImageDoesNotExist
+		}
+		mrhi.micHelper.SetImageStatus(micObj, imageState.Image, micImageStatus)
+	}
+
+	return mrhi.client.Status().Patch(ctx, micObj, patchFrom)
+}
+
+func (mrhi *micReconcilerHelperImpl) processImagesSpecs(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig, pullPods []v1.Pod) error {
+	errs := make([]error, len(micObj.Spec.Images))
+	for _, imageSpec := range micObj.Spec.Images {
+		imageState := mrhi.micHelper.GetImageState(micObj, imageSpec.Image)
+
+		// image State does not exists either new image or pull pod is still running
+		switch imageState {
+		case "":
+			// image State is not set: either new image or pull pod is still running
+			if mrhi.podHelper.getPullPodForImage(pullPods, imageSpec.Image) == nil {
+				// no pull pod- create it, otherwise we wait for it to finish
+				err := mrhi.podHelper.createPullPod(ctx, &imageSpec, micObj)
+				errs = append(errs, err)
+			}
+		case kmmv1beta1.ImageDoesNotExist:
+			if imageSpec.Build == nil && imageSpec.Sign == nil {
+				break
+			}
+			fallthrough
+		case kmmv1beta1.ImageNeedsBuilding:
+			// image needs to be built - patching MBSC
+			mbscObj := &kmmv1beta1.ModuleBuildSignConfig{
+				ObjectMeta: metav1.ObjectMeta{Name: micObj.Name, Namespace: micObj.Namespace},
+			}
+			_, err := controllerutil.CreateOrPatch(ctx, mrhi.client, mbscObj, func() error {
+				mrhi.mbscHelper.SetModuleImageSpec(mbscObj, &imageSpec)
+				mbscObj.Spec.ImageRepoSecret = micObj.Spec.ImageRepoSecret
+				return controllerutil.SetOwnerReference(micObj, mbscObj, mrhi.scheme)
+			})
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+//go:generate mockgen -source=mic_reconciler.go -package=controllers -destination=mock_mic_reconciler.go pullPodManager
+type pullPodManager interface {
+	listImagesPullPods(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig) ([]v1.Pod, error)
+	deletePod(ctx context.Context, pod *v1.Pod) error
+	createPullPod(ctx context.Context, imageSpec *kmmv1beta1.ModuleImageSpec, micObj *kmmv1beta1.ModuleImagesConfig) error
+	getPullPodForImage(pods []v1.Pod, image string) *v1.Pod
+}
```

Comment:
How about we create a public interface for pods instead of another reconciler helper API?
* listImagesPullPods only needs mic for name and namespace
* createPullPod can get parameters like name,ns, image (string) instead of MIC/MICSpec

Reply:
This API is targeted specifically for MIC.  Creating it as a public interface which is used only by one controller does not make much of a difference.We have another helper pod interface , in NMC controller. We might consider unifying them , but in any case it should be in a different PR.

Reply:
Yes, I do think we should unify them and create a public interface for them. I will open a task for it.